File: src/lib/project.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  EffortLevel,
  Project,
  ProjectMemoryEntry,
  ProjectSettingsData,
  MemoryAgentConfig,
  MemoryAgentTestResult,
  Repository,
  WorkflowOverride,
} from "../types";

// ── Project records ─────────────────────────────────────────────────────

export async function getProjects(): Promise<Project[]> {
  return invoke<Project[]>("get_projects");
}

/** Create the demo project. The backend pins a fixed id and discards the
 *  insert result, so a second call resolves with the same record whether or
 *  not a row was written — treat the result as the intended shape, not as
 *  proof of what is stored. */
export async function seedSampleProject(): Promise<Project> {
  return invoke<Project>("seed_sample_project");
}

export async function getRepositoriesForProject(projectId: string): Promise<Repository[]> {
  return invoke<Repository[]>("get_repositories_for_project", { projectId });
}

/** Mirrors the Rust `ProjectConfig`. Sent whole on every update, so a caller
 *  that drops `repos` drops the project's repositories. */
export interface ProjectConfigInput {
  name: string;
  compute_type: string;
  remote_host: string | null;
  repos: Array<{ repo_path: string; provider_id: string }>;
}

export async function updateProject(id: string, config: ProjectConfigInput): Promise<void> {
  return invoke<void>("update_project", { id, config });
}

/** Delete the project and its cloned workspace. */
export async function deleteProject(id: string): Promise<void> {
  return invoke<void>("delete_project", { id });
}

/** Uncommitted / unpushed state of the named repos. Mirrors the Rust
 *  `RepoDirtyStatus`. */
export interface RepoDirtyStatus {
  repo_path: string;
  has_uncommitted: boolean;
  has_unpushed: boolean;
}

export async function checkReposDirty(
  projectId: string,
  repoPaths: string[],
): Promise<RepoDirtyStatus[]> {
  return invoke<RepoDirtyStatus[]>("check_repos_dirty", { projectId, repoPaths });
}

export interface WorktreeInfo {
  path: string;
  branch: string | null;
  is_locked: boolean;
}

/** Mirrors the Rust `RepoHealthStatus`. */
export interface RepoHealthStatus {
  repo_path: string;
  is_cloned: boolean;
  head_branch: string | null;
  worktrees: WorktreeInfo[];
  has_uncommitted: boolean;
  has_unpushed: boolean;
}

export async function getWorkspaceHealth(projectId: string): Promise<RepoHealthStatus[]> {
  return invoke<RepoHealthStatus[]>("get_workspace_health", { projectId });
}

/** The project's stored settings, or `null` before its first save. */
export async function getProposedStrategy(
  projectId: string,
): Promise<ProjectSettingsData | null> {
  return invoke<ProjectSettingsData | null>("get_proposed_strategy", { projectId });
}

// ── Project memory ──────────────────────────────────────────────────────

export async function listProjectMemory(projectId: string): Promise<ProjectMemoryEntry[]> {
  return invoke<ProjectMemoryEntry[]>("project_memory_list", { projectId });
}

export async function upsertProjectMemory(
  projectId: string,
  key: string,
  value: string,
  source: 'agent' | 'human',
  id?: string | null,
): Promise<void> {
  return invoke<void>("project_memory_upsert", {
    id: id || null,
    projectId,
    key,
    value,
    source,
  });
}

export async function deleteProjectMemory(id: string): Promise<void> {
  return invoke<void>("project_memory_delete", { id });
}

// ── Memory agent (global) ──────────────────────────────────────────────

export async function getMemoryAgentConfig(): Promise<MemoryAgentConfig> {
  return invoke<MemoryAgentConfig>("memory_agent_config_get");
}

/** Persist config. `apiKey`: `undefined` keeps the stored key, `''` clears it,
 * a non-empty string stores a new key. */
export async function setMemoryAgentConfig(
  config: MemoryAgentConfig,
  apiKey?: string,
): Promise<void> {
  return invoke<void>("memory_agent_config_set", {
    config,
    apiKey: apiKey === undefined ? null : apiKey,
  });
}

export async function testMemoryAgentConnection(
  config: MemoryAgentConfig,
  apiKey?: string,
): Promise<MemoryAgentTestResult> {
  return invoke<MemoryAgentTestResult>("memory_agent_test_connection", {
    config,
    apiKey: apiKey === undefined ? null : apiKey,
  });
}

/** List models available at an endpoint (OpenAI `/models`, falling back to
 * Ollama `/api/tags`). */
export async function listMemoryAgentModels(
  endpoint: string,
  apiKey?: string,
): Promise<string[]> {
  return invoke<string[]>("memory_agent_list_models", {
    endpoint,
    apiKey: apiKey === undefined ? null : apiKey,
  });
}

// ── Workflow / step overrides ──────────────────────────────────────────

/** Every override configured for a project, workflow-level and step-level.
 *  Anything inheriting has no row and is absent from the list. */
export async function getWorkflowOverrides(projectId: string): Promise<WorkflowOverride[]> {
  return invoke<WorkflowOverride[]>("get_workflow_overrides", { projectId });
}

/**
 * Upsert one project-scoped override. `stepId === null` is the
 * workflow-level row ("applies to all steps"); a step id targets one step.
 * Each field is independently `null` = "inherit that one field"; all three
 * `null` clears the row entirely (the repo deletes it).
 */
export async function setWorkflowOverride(input: {
  projectId: string;
  workflowId: string;
  stepId: string | null;
  agentKind: string | null;
  model: string | null;
  effort: EffortLevel | null;
}): Promise<void> {
  await invoke<void>("set_workflow_override", {
    projectId: input.projectId,
    workflowId: input.workflowId,
    stepId: input.stepId,
    agentKind: input.agentKind,
    model: input.model,
    effort: input.effort,
  });
}

// ── Configuration-time command probe (HB6) ─────────────────────────────────

/** Which project setting a probed command came from. */
export type ProbedCommandSource = 'prepare' | 'test' | 'harness';

/** One binary a configured command names, and whether the machine found it.
 *  A binary the engine deliberately skipped (a shell builtin, a `$(…)`
 *  substitution, a glob) is absent from the list rather than reported as
 *  healthy — the panel claims exactly what was checked. */
export interface ProbedBinary {
  name: string;
  resolved: boolean;
}

export interface ProbedCommand {
  source: ProbedCommandSource;
  /** The `harnesses` key, for `source === 'harness'`. */
  harness: string | null;
  command: string;
  binaries: ProbedBinary[];
}

export interface CommandProbeReport {
  /** The machine that was actually asked — the *project's*, not the laptop's. */
  machine: string;
  commands: ProbedCommand[];
  /** The engine's own `PreflightVerdict::detail` — the same string a blocked
   *  launch terminates with, carrying the `bash -l -i -c` reproduce line.
   *  Rendered verbatim so the panel and the launch cannot drift apart. */
  detail: string | null;
  /** The engine's fresh-checkout / watch-mode remediation, likewise verbatim. */
  guidance: string;
  /** Whether this verdict would stop a launch. Reported, never enforced: a
   *  save is not gated on a probe. */
  blocks_launch: boolean;
}

/**
 * Probe the project's configured commands on **the project's own machine**.
 *
 * The commands are sent as they stand in the form, not read back from the DB:
 * the point is to answer for the command the user just typed. Which machine
 * gets asked is decided backend-side from the project's compute type, so the
 * panel cannot accidentally report the laptop's PATH for a remote project.
 */
export async function probeProjectCommands(input: {
  projectId: string;
  prepareCommand: string | null;
  testCommand: string | null;
  harnesses: Record<string, string> | null;
}): Promise<CommandProbeReport> {
  return invoke<CommandProbeReport>('probe_project_commands', {
    projectId: input.projectId,
    draft: {
      prepare_command: input.prepareCommand || null,
      test_command: input.testCommand || null,
      harnesses: input.harnesses && Object.keys(input.harnesses).length > 0 ? input.harnesses : null,
    },
  });
}

/**
 * Partial project-settings input. Any field left `undefined` is filled from
 * the existing DB record (or a sensible default). This prevents the
 * partial-save data-loss bug where a caller that omits a field would
 * accidentally `INSERT OR REPLACE` it to NULL.
 */
export interface ProjectSettingsInput {
  default_branch?: string;
  branch_prefix?: string;
  test_command?: string | null;
  build_command?: string | null;
  coverage_command?: string | null;
  conventions_file?: string | null;
  pr_template?: string | null;
  harnesses?: Record<string, string> | null;
  /** The user's ordered selection of which harnesses gate validation — tier 2
   *  of the engine's harness resolution chain. Order is the user's (cheap
   *  gates first) and has to be stored, because `harnesses` is a map with no
   *  order to inherit. `null`/empty = no selection, which resolves exactly as
   *  it does today. */
  validation_gates?: string[] | null;
  prepare_command?: string | null;
  extra_writable_paths?: string[] | null;
  conflict_policy?: string;
  feature_lifecycle?: string;
  default_agent_kind?: string | null;
  default_model?: string | null;
  default_effort?: EffortLevel | null;
  default_loop_iterations?: number | null;
  default_max_budget_usd?: number | null;
  artifact_subdir?: string;
  commit_artifacts?: boolean;
}

/**
 * Read existing DB settings, overlay the caller's partial input, and write
 * back the complete merged record.  Any field omitted from `input` (i.e.
 * left `undefined`) keeps whatever is already in the database, so a save
 * call that only touches a few form fields can never NULL out the rest.
 */
export async function saveProjectSettings(
  projectId: string,
  input: ProjectSettingsInput,
): Promise<void> {
  const existing = await getProposedStrategy(projectId);
  const baseWs = existing?.worktree_strategy;

  const merged = {
    project_id: projectId,
    worktree_strategy: {
      default_branch:
        input.default_branch ?? baseWs?.default_branch ?? "main",
      branch_prefix:
        input.branch_prefix ?? baseWs?.branch_prefix ?? "demeteo/features/",
      test_command:
        input.test_command !== undefined
          ? input.test_command
          : (baseWs?.test_command ?? null),
      build_command:
        input.build_command !== undefined
          ? input.build_command
          : (baseWs?.build_command ?? null),
      coverage_command:
        input.coverage_command !== undefined
          ? input.coverage_command
          : (baseWs?.coverage_command ?? null),
      conventions_file:
        input.conventions_file !== undefined
          ? input.conventions_file
          : (baseWs?.conventions_file ?? null),
      pr_template:
        input.pr_template !== undefined
          ? input.pr_template
          : (baseWs?.pr_template ?? null),
      harnesses:
        input.harnesses !== undefined
          ? (Object.keys(input.harnesses ?? {}).length > 0
              ? input.harnesses
              : null)
          : (baseWs?.harnesses ?? null),
      // Dropped rather than written when the selection is empty: the backend
      // stores the bare `harnesses` map in that case, so a project that never
      // ticks a gate keeps writing byte-identical rows.
      validation_gates:
        input.validation_gates !== undefined
          ? (input.validation_gates && input.validation_gates.length > 0
              ? input.validation_gates
              : null)
          : (baseWs?.validation_gates ?? null),
      prepare_command:
        input.prepare_command !== undefined
          ? input.prepare_command
          : (baseWs?.prepare_command ?? null),
      extra_writable_paths:
        input.extra_writable_paths !== undefined
          ? (Array.isArray(input.extra_writable_paths)
              ? input.extra_writable_paths
              : [])
          : (baseWs?.extra_writable_paths ?? []),
    },
    conflict_policy:
      input.conflict_policy ?? existing?.conflict_policy ?? "always_gate",
    feature_lifecycle:
      input.feature_lifecycle ?? existing?.feature_lifecycle ?? "archive",
    default_agent_kind:
      input.default_agent_kind !== undefined
        ? input.default_agent_kind
        : (existing?.default_agent_kind ?? null),
    default_model:
      input.default_model !== undefined
        ? input.default_model
        : (existing?.default_model ?? null),
    // Omitted by every caller that doesn't own the Strategy tab, so it has to
    // be carried across from the stored record — otherwise saving any other
    // setting would silently drop the project's default effort.
    default_effort:
      input.default_effort !== undefined
        ? input.default_effort
        : (existing?.default_effort ?? null),
    default_loop_iterations:
      input.default_loop_iterations !== undefined
        ? input.default_loop_iterations
        : (existing?.default_loop_iterations ?? null),
    default_max_budget_usd:
      input.default_max_budget_usd !== undefined
        ? input.default_max_budget_usd
        : (existing?.default_max_budget_usd ?? null),
    artifact_subdir:
      input.artifact_subdir ?? existing?.artifact_subdir ?? "artifacts/",
    commit_artifacts:
      input.commit_artifacts ?? existing?.commit_artifacts ?? false,
  };

  await invoke("save_project_settings", { projectId, settings: merged });
}

File: crates/demeteo-core/src/domain/models/project.rs
use crate::domain::ids::{MachineId, ProjectId, ProviderId, RepositoryId, WorkflowId};
use crate::domain::models::EffortLevel;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Project {
    pub id: ProjectId,
    pub name: String,
    pub compute_type: String, // 'local' | 'remote'
    pub remote_host: Option<MachineId>,
    pub status: String,
    pub nodes: i32,
    pub spend: f64,
    #[serde(default)]
    pub tokens: i64,
    pub created_at: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Repository {
    pub id: RepositoryId,
    pub project_id: ProjectId,
    pub provider_id: ProviderId,
    pub repo_path: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct WorktreeStrategy {
    pub default_branch: String,
    pub branch_prefix: String,
    #[serde(default)]
    pub test_command: Option<String>,
    #[serde(default)]
    pub build_command: Option<String>,
    #[serde(default)]
    pub coverage_command: Option<String>,
    #[serde(default)]
    pub conventions_file: Option<String>,
    pub pr_template: Option<String>,
    #[serde(default)]
    pub harnesses: Option<HashMap<String, String>>,
    /// The user's **ordered selection of which harnesses gate validation** —
    /// tier 2 of [`resolve_harnesses`](crate::domain::verifier::resolve_harnesses).
    ///
    /// It exists because `harnesses` was otherwise dead config: the only thing
    /// that selected an entry was a per-step `verifier.harness_names`, and all
    /// seven shipped starters declare none — so a user could add
    /// `lint → npm run lint`, see it accepted, and nothing would ever run it
    /// short of forking a starter. Ticking it here gates every workflow that
    /// declares nothing of its own.
    ///
    /// Ordered because `harnesses` is a `HashMap` and therefore has no order to
    /// inherit, while the order is the user's (cheap gates first — lint before
    /// integration). Names that no longer exist in `harnesses` are ignored at
    /// resolution rather than erroring.
    ///
    /// `None`/empty = no selection, which resolves exactly as it does today.
    /// Persisted inside the `harnesses` column — see
    /// `adapters/database/repos/project.rs`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub validation_gates: Option<Vec<String>>,
    /// Optional shell command run inside each subtask worktree before the
    /// verifier's harness command (`npm ci`, `cargo fetch`, `prisma
    /// generate`, a DB migration, …). Runs after write permissions are
    /// restored and after `provision_subtask_worktree`'s dependency-cache
    /// symlinking, so it only needs to handle what symlinking a prior
    /// install can't — codegen, migrations, freshly-added dependencies.
    /// `None` (default) skips this step entirely.
    #[serde(default)]
    pub prepare_command: Option<String>,
    /// Project-wide writability exceptions, applied on top of the
    /// capability-driven chmod fence. Repo-relative paths the agent may
    /// write to even when the step's capability (`ReadOnly`,
    /// `Artifacts`, `Verify`) would otherwise fence them. Designed for
    /// tool side-effects that aren't source or artifacts — e.g.
    /// `target/` for `cargo test`, `node_modules/` for `npm test`,
    /// `.venv/` for `pytest`. Each entry must be a relative path
    /// inside the worktree; `..` is rejected to prevent escape.
    /// Stays empty for `Implement` capability (which is already
    /// fully writable). See scope adapter `derive_writable_paths_for_scope`.
    #[serde(default)]
    pub extra_writable_paths: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ProjectSettings {
    pub project_id: ProjectId,
    pub worktree_strategy: WorktreeStrategy,
    pub conflict_policy: String,
    pub feature_lifecycle: String,
    #[serde(default)]
    pub default_agent_kind: Option<String>,
    #[serde(default)]
    pub default_model: Option<String>,
    /// Project-wide default effort for every step of every run. `None` = fall
    /// through to [`EffortLevel::DEFAULT`] (high). The lowest tier of the
    /// effort resolution chain; a workflow, a launch override, or a project
    /// workflow override all beat it. See migration V29.
    #[serde(default)]
    pub default_effort: Option<EffortLevel>,
    /// The Workflow a new Feature starts with unless the launcher picks
    /// another. `None` = **unset**, not "no workflow": the launch path then
    /// falls back to the first entry of the workflow list, which is what
    /// every project did implicitly before this field existed.
    ///
    /// Read at launch time only. Unlike `default_effort` and `default_model`
    /// it feeds no resolution chain during a run, and a workflow deleted
    /// after being chosen leaves this id dangling by design — the caller
    /// resolves it against the workflow list and treats an unresolvable id
    /// as unset. See migration V40 for why there is no foreign key.
    #[serde(default)]
    pub default_workflow_id: Option<String>,
    /// Project-level default loop iteration budget for `on_failure` retry
    /// loops. `None` = use the engine default (3). Overridable per run via
    /// `Feature::loop_iterations`. See migration V13.
    #[serde(default)]
    pub default_loop_iterations: Option<u32>,
    /// Project-level default per-turn dollar budget (`--max-budget-usd`).
    /// `None` = use the engine default
    /// ([`crate::domain::agent_session::budget::DEFAULT_MAX_BUDGET_USD`]).
    /// Overridable per run via
    /// `Feature::max_budget_usd`. See migration V30.
    #[serde(default)]
    pub default_max_budget_usd: Option<f64>,
    /// Repo-relative folder where agents write their reports
    /// (`research-report.md`, `critic-review.md`, …). The orchestrator
    /// injects `{{report_dir}}` (alias `{{artifact_dir}}`, kept for
    /// back-compat with older workflows) into every step's prompt and
    /// excludes this folder from `commit_worktree_changes` unless
    /// `commit_artifacts` is true. Default: `"artifacts/"`.
    /// See migration V12 and AGENTS.md §6.
    ///
    /// Note for `wf-starter-docs-update`: the docs-update workflow's
    /// `s-draft` and `s-polish` steps are explicitly told (in their
    /// `prompt_template`) to write the real doc body at the path the
    /// survey/gate approved (typically under `docs/`) and to use
    /// `{{report_dir}}` ONLY for the short change-summary report.
    /// That separation is what keeps a "create a new doc explaining
    /// feature X" feature from silently landing its body under
    /// `artifacts/s-draft.md` (which `commit_artifacts=false` would
    /// keep off the branch). The StartFeatureModal's advanced
    /// section surfaces this toggle so users can opt a single
    /// docs-update feature into committing its reports alongside the
    /// new doc.
    #[serde(default = "default_artifact_subdir")]
    pub artifact_subdir: String,
    /// When false (default), the orchestrator's
    /// `commit_worktree_changes` runs `git add -A -- ':!<artifact_subdir>'`
    /// so the reports stay in the worktree as untracked files instead of
    /// being committed into the feature branch. The reports' content is
    /// still captured into the `FsArtifactStore` for the UI.
    /// Per-feature override lives on `Feature::commit_artifacts`.
    ///
    /// Note for `wf-starter-docs-update`: leave this `false` (the
    /// default) for the "create a new doc" case so the new doc body
    /// at its real `docs/...` path lands on the branch while the
    /// `artifacts/s-draft.md` summary stays out. Flip it to `true`
    /// (per-feature via the StartFeatureModal advanced section) if
    /// the user wants both the doc and the change-summary report in
    /// the same commit.
    ///
    /// Note for remote runner runs (C4.4): the default `false` used to be
    /// a silent data-loss trap for remote features — the PR was the only
    /// channel back to the laptop, so uncommitted reports were invisible.
    /// That no longer holds: C4.2's shadow mirror pulls each remote step's
    /// declared artifacts into the laptop `FsArtifactStore` and `RunView`
    /// renders them via the return inbox's "View feature". So `false`
    /// stays the clean default for remote too (reports reach the laptop
    /// through the mirror, not the branch); `true` is now a deliberate
    /// "also commit them to the PR" opt-in, not the only way to see them.
    #[serde(default)]
    pub commit_artifacts: bool,
    /// The command a reviewing step should start from, as the user wrote it.
    /// Bound to `{{review_entrypoint}}` and carried into the prompt untouched —
    /// [`review_entrypoint`](crate::domain::review_entrypoint) holds why it is
    /// never wrapped, and why `None` and `""` are the same answer. See
    /// migration V42.
    #[serde(default)]
    pub review_entrypoint: Option<String>,
    /// The harness a conflict-resolution turn should run under, outranking the
    /// run's own launch pin for that turn alone.
    /// [`sync_resolver`](crate::domain::sync_resolver) holds why a role gets a
    /// tier above the run, and what `None` falls through to. See migration V44.
    #[serde(default)]
    pub sync_resolver_agent_kind: Option<String>,
    /// The model for that turn, inherited independently of the harness.
    #[serde(default)]
    pub sync_resolver_model: Option<String>,
    /// The reasoning effort for that turn. Clamped per harness at spawn, so a
    /// level the chosen harness does not offer is never emitted.
    #[serde(default)]
    pub sync_resolver_effort: Option<EffortLevel>,
    /// Whether a resolved sync waits for a human before it is published.
    /// `None` is "no opinion", which is not `false` —
    /// [`publish_policy`](crate::domain::sync_session::publish_policy) holds
    /// what it resolves to and why the setting can only turn review off. See
    /// migration V45.
    #[serde(default)]
    pub sync_review_before_push: Option<bool>,
}

fn default_artifact_subdir() -> String {
    "artifacts/".to_string()
}

/// A subset of `ProjectSettings` writable by a non-UI external caller (the
/// `agent_surface` seam) — run shape only, never a spend or safety boundary.
///
/// Excluded on purpose, and not by omission:
/// - `default_max_budget_usd`, `default_loop_iterations`: per-run cost
///   ceilings. A caller that could raise either is raising the dollar/retry
///   bound on every run it subsequently causes — the permission split is
///   meaningless if the same call can both start a run and widen its own
///   leash.
/// - `sync_review_before_push`: switches off the human review gate before a
///   push. Same shape of boundary as budget: a caller must not be able to
///   remove the check on its own output.
/// - `worktree_strategy`, `feature_lifecycle`: project topology, not run
///   shape — set once when the project is configured, not per launch.
///
/// Rejected: sharing `ProjectSettings` whole with the UI's save path
/// (`save_project_settings`). The UI is a human looking at a labelled
/// control; this caller is not. A shared path would grant budget and
/// review-bypass changes under a call that reads like "save my
/// preferences."
///
/// A field added to `ProjectSettings` later is unreachable here until
/// someone deliberately adds it below, having decided it is run-shape and
/// not a boundary — that decision is the point of this being a type.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunShapePatch {
    #[serde(default)]
    pub default_agent_kind: Option<String>,
    #[serde(default)]
    pub default_model: Option<String>,
    #[serde(default)]
    pub default_effort: Option<EffortLevel>,
    #[serde(default)]
    pub default_workflow_id: Option<String>,
    #[serde(default = "default_artifact_subdir")]
    pub artifact_subdir: String,
    #[serde(default)]
    pub commit_artifacts: bool,
    #[serde(default)]
    pub sync_resolver_agent_kind: Option<String>,
    #[serde(default)]
    pub sync_resolver_model: Option<String>,
    #[serde(default)]
    pub sync_resolver_effort: Option<EffortLevel>,
}

/// Why a [`RunShapePatch`] was refused before it reached storage.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum RunShapePatchError {
    #[error(
        "artifact_subdir must be a relative path of letters, digits, spaces and `_-./`, \
         with no `.` or `..` segments"
    )]
    ArtifactSubdir,
    #[error("{0} must not start with `-` or contain whitespace or control characters")]
    Identifier(&'static str),
}

impl RunShapePatch {
    /// A typed patch is only as safe as its field types, and these are all
    /// `String`. `artifact_subdir` ends up inside a `git add` pathspec and an
    /// artifact-scope path; the agent/model/workflow fields end up as argv
    /// values of a spawned harness and as lookup keys. None of them is a shell
    /// sink today (the sink escapes — see `resolve_add_exclusions`), so this is
    /// the second fence, not the first: an external caller gets a refusal
    /// instead of a value that later runs are left to survive.
    pub fn validate(&self) -> Result<(), RunShapePatchError> {
        let subdir =
            crate::domain::staged_deliverable::normalize_artifact_subdir(&self.artifact_subdir);
        let subdir_ok = subdir.is_empty()
            || (!subdir.starts_with('/')
                && subdir.chars().all(|c| {
                    c.is_ascii_alphanumeric() || matches!(c, ' ' | '_' | '-' | '.' | '/')
                })
                && subdir
                    .split('/')
                    .all(|seg| !seg.is_empty() && seg != "." && seg != ".."));
        if !subdir_ok {
            return Err(RunShapePatchError::ArtifactSubdir);
        }

        let identifiers = [
            ("default_agent_kind", &self.default_agent_kind),
            ("default_model", &self.default_model),
            ("default_workflow_id", &self.default_workflow_id),
            ("sync_resolver_agent_kind", &self.sync_resolver_agent_kind),
            ("sync_resolver_model", &self.sync_resolver_model),
        ];
        for (field, value) in identifiers {
            if let Some(v) = value {
                if v.starts_with('-') || v.chars().any(|c| c.is_whitespace() || c.is_control()) {
                    return Err(RunShapePatchError::Identifier(field));
                }
            }
        }
        Ok(())
    }
}

/// Copies `patch`'s nine fields onto `settings`, returning every other
/// field unchanged. See [`RunShapePatch`] for which fields those are and why.
pub fn apply_run_shape_patch(
    mut settings: ProjectSettings,
    patch: RunShapePatch,
) -> ProjectSettings {
    settings.default_agent_kind = patch.default_agent_kind;
    settings.default_model = patch.default_model;
    settings.default_effort = patch.default_effort;
    settings.default_workflow_id = patch.default_workflow_id;
    settings.artifact_subdir = patch.artifact_subdir;
    settings.commit_artifacts = patch.commit_artifacts;
    settings.sync_resolver_agent_kind = patch.sync_resolver_agent_kind;
    settings.sync_resolver_model = patch.sync_resolver_model;
    settings.sync_resolver_effort = patch.sync_resolver_effort;
    settings
}

/// A project-scoped override of the coding agent ("harness") and/or model
/// for a (global) workflow — either the whole workflow or a single step.
/// Persisted in `project_workflow_overrides` (migrations V14 / V15).
///
/// Scope is set by `step_id`:
///   - `None` → workflow-level. At feature start it overlays the project
///     defaults (`ProjectSettings::default_agent_kind` / `default_model`) for
///     this workflow only.
///   - `Some(step_id)` → step-level. It is baked onto the matching
///     `StepConfig`, so it beats the workflow author's value for that step.
///
/// In all cases it still loses to a run-time override on the feature row —
/// the feature-wide pair chosen at launch, and above it the per-step
/// [`StepOverride`](crate::domain::models::StepOverride), which is writable
/// for as long as the run is alive. `None` on a field = inherit for that field.
/// See `resolve_execution_context`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ProjectWorkflowOverride {
    pub project_id: ProjectId,
    pub workflow_id: WorkflowId,
    /// `None` = workflow-level (stored as `''`); `Some` targets one step.
    #[serde(default)]
    pub step_id: Option<String>,
    #[serde(default)]
    pub agent_kind: Option<String>,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub effort: Option<EffortLevel>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublishOptions {
    #[serde(default)]
    pub draft: bool,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub body: Option<String>,
    #[serde(default)]
    pub target_branch: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MrInfo {
    pub url: String,
    pub state: String,
    pub number: u64,
    pub provider_kind: String,
    pub provider_host: String,
}

#[cfg(test)]
#[path = "../../../tests/domain/models/project.rs"]
mod tests;
